File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/felles/ÅrstallOgKvartal.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import lombok.Data;
import org.jetbrains.annotations.NotNull;

@Data
public class ÅrstallOgKvartal implements Comparable<ÅrstallOgKvartal> {

    // TODO: Disse verdiene må oppdateres hver gang vi publiserer statistikk for nytt kvartal.
    // Det bør automatiseres bort på et vis.
    public static ÅrstallOgKvartal SISTE_PUBLISERTE_KVARTAL = new ÅrstallOgKvartal(2022, 2);

    private int årstall;
    private int kvartal;

    public static ÅrstallOgKvartal sisteKvartalMinus(int n) {
        return SISTE_PUBLISERTE_KVARTAL.minusKvartaler(n);
    }

    public ÅrstallOgKvartal(int årstall, int kvartal) {
        if (kvartal > 4 || kvartal < 1) {
            throw new IllegalArgumentException("Kvartal må være 1, 2, 3 eller 4");
        }
        this.årstall = årstall;
        this.kvartal = kvartal;
    }

    public ÅrstallOgKvartal minusKvartaler(int antallKvartaler) {
        if (antallKvartaler < 0) {
            return plussKvartaler(-antallKvartaler);
        }
        ÅrstallOgKvartal årstallOgKvartal = new ÅrstallOgKvartal(årstall, kvartal);
        for (int i = 0; i < antallKvartaler; i++) {
            årstallOgKvartal = årstallOgKvartal.forrigeKvartal();
        }
        return årstallOgKvartal;
    }

    public ÅrstallOgKvartal minusEttÅr() {
        return new ÅrstallOgKvartal(this.årstall - 1, this.kvartal);
    }

    public static List<ÅrstallOgKvartal> sisteFireKvartaler(ÅrstallOgKvartal årstallOgKvartal) {
        return IntStream.range(0, 4)
                .mapToObj(årstallOgKvartal::minusKvartaler)
                .collect(Collectors.toList());
    }

    public ÅrstallOgKvartal plussKvartaler(int antallKvartaler) {
        if (antallKvartaler < 0) {
            return minusKvartaler(-antallKvartaler);
        }
        ÅrstallOgKvartal årstallOgKvartal = new ÅrstallOgKvartal(årstall, kvartal);
        for (int i = 0; i < antallKvartaler; i++) {
            årstallOgKvartal = årstallOgKvartal.nesteKvartal();
        }
        return årstallOgKvartal;
    }

    public static List<ÅrstallOgKvartal> range(ÅrstallOgKvartal fra, ÅrstallOgKvartal til) {
        List<ÅrstallOgKvartal> årstallOgKvartal = new ArrayList<>();
        for (ÅrstallOgKvartal i = fra; i.compareTo(til) <= 0; i = i.plussKvartaler(1)) {
            årstallOgKvartal.add(i);
        }
        return årstallOgKvartal;
    }

    private ÅrstallOgKvartal forrigeKvartal() {
        if (kvartal == 1) {
            return new ÅrstallOgKvartal(årstall - 1, 4);
        } else {
            return new ÅrstallOgKvartal(årstall, kvartal - 1);
        }
    }

    private ÅrstallOgKvartal nesteKvartal() {
        if (kvartal == 4) {
            return new ÅrstallOgKvartal(årstall + 1, 1);
        } else {
            return new ÅrstallOgKvartal(årstall, kvartal + 1);
        }
    }

    @Override
    public int compareTo(@NotNull ÅrstallOgKvartal årstallOgKvartal) {
        return Comparator.comparing(ÅrstallOgKvartal::getÅrstall)
                .thenComparing(ÅrstallOgKvartal::getKvartal)
                .compare(this, årstallOgKvartal);
    }

    @Override
    public String toString() {
        return kvartal + ". kvartal " + årstall;
    }
}

File: src/main/kotlin/no/nav/arbeidsgiver/sykefravarsstatistikk/api/applikasjon/fellesdomene/ÅrstallOgKvartal.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.fellesdomene

data class ÅrstallOgKvartal(
    val årstall: Int,
    val kvartal: Int
) : Comparable<ÅrstallOgKvartal> {

    init {
        require(!(kvartal > 4 || kvartal < 1)) { "Kvartal må være 1, 2, 3 eller 4" }
    }

    fun minusKvartaler(antallKvartaler: Int): ÅrstallOgKvartal {
        if (antallKvartaler < 0) {
            return plussKvartaler(-antallKvartaler)
        }
        var årstallOgKvartal = ÅrstallOgKvartal(årstall, kvartal)
        for (i in 0 until antallKvartaler) {
            årstallOgKvartal = årstallOgKvartal.forrigeKvartal()
        }
        return årstallOgKvartal
    }

    fun minusEttÅr(): ÅrstallOgKvartal {
        return ÅrstallOgKvartal(årstall - 1, kvartal)
    }

    fun plussKvartaler(antallKvartaler: Int): ÅrstallOgKvartal {
        if (antallKvartaler < 0) {
            return minusKvartaler(-antallKvartaler)
        }
        var årstallOgKvartal = ÅrstallOgKvartal(årstall, kvartal)
        for (i in 0 until antallKvartaler) {
            årstallOgKvartal = årstallOgKvartal.nesteKvartal()
        }
        return årstallOgKvartal
    }

    private fun forrigeKvartal(): ÅrstallOgKvartal {
        return if (kvartal == 1) {
            ÅrstallOgKvartal(årstall - 1, 4)
        } else {
            ÅrstallOgKvartal(årstall, kvartal - 1)
        }
    }

    private fun nesteKvartal(): ÅrstallOgKvartal {
        return if (kvartal == 4) {
            ÅrstallOgKvartal(årstall + 1, 1)
        } else {
            ÅrstallOgKvartal(årstall, kvartal + 1)
        }
    }

    override fun compareTo(other: ÅrstallOgKvartal): Int {
        return Comparator.comparing<ÅrstallOgKvartal, Int> { it.årstall }
            .thenComparing<Int> { it.kvartal }
            .compare(this, other)
    }

    override fun toString(): String {
        return "$kvartal. kvartal $årstall"
    }

    companion object {

        fun sisteFireKvartaler(årstallOgKvartal: ÅrstallOgKvartal): List<ÅrstallOgKvartal> =
            (0..3).map(årstallOgKvartal::minusKvartaler).sorted()


        fun range(fra: ÅrstallOgKvartal, til: ÅrstallOgKvartal): List<ÅrstallOgKvartal> {
            val årstallOgKvartal: MutableList<ÅrstallOgKvartal> = ArrayList()
            var i = fra
            while (i <= til) {
                årstallOgKvartal.add(i)
                i = i.plussKvartaler(1)
            }
            return årstallOgKvartal
        }
    }
}
